File: leetcode/python/000079.word-search.py
class Solution:
    # Time complexity: O(n * m * 4^L) -> n: rows, m: cols, L: len(word)
    # Space complexity: O(L) -> Visited set and recursion depth can only be as long as L
    def exist(self, board: list[list[str]], word: str) -> bool:
        ROWS, COLS = len(board), len(board[0])
        visited = set()

        def dfs(r: int, c: int, i: int):
            if i == len(word):
                return True

            if (
                r < 0
                or r >= ROWS
                or c < 0
                or c >= COLS
                or (r, c) in visited
                or board[r][c] != word[i]
            ):
                return False

            visited.add((r, c))

            res = (
                dfs(r + 1, c, i + 1)
                or dfs(r - 1, c, i + 1)
                or dfs(r, c + 1, i + 1)
                or dfs(r, c - 1, i + 1)
            )

            visited.remove((r, c))

            return res

        for i in range(ROWS):
            for j in range(COLS):
                if dfs(i, j, 0):
                    return True

        return False

File: leetcode/javascript/000079.word-search.js
/**
 * @param {character[][]} board
 * @param {string} word
 * @return {boolean}
 * Time complexity: O(m * n * 4^L) -> 4 choices per step
 * Space complexity: O(L)
 */
var exist = function (board, word) {
  function key(r, c) {
    return `${r},${c}`;
  }

  const rows = board.length;
  const cols = board[0].length;
  const visited = new Set();

  function dfs(i, r, c) {
    if (
      r < 0 || c < 0 || r >= rows || c >= cols || visited.has(key(r, c)) ||
      board[r][c] !== word[i]
    ) {
      return false;
    }

    if (i === word.length - 1) {
      return true;
    }

    visited.add(key(r, c));
    const result = dfs(i + 1, r - 1, c) || dfs(i + 1, r + 1, c) ||
      dfs(i + 1, r, c - 1) || dfs(i + 1, r, c + 1);
    visited.delete(key(r, c));

    return result;
  }

  for (let r = 0; r < rows; r++) {
    for (let c = 0; c < cols; c++) {
      if (dfs(0, r, c)) {
        return true;
      }
    }
  }

  return false;
};
